projects/feature-activation/0005-feature-activation-for-transactions.md

```diff
@@ -0,0 +1,196 @@
+    """
+    from hathor.transaction import Block, Transaction
+    if isinstance(self, Block):
+        return
+    assert isinstance(self, Transaction)
+    assert self.storage is not None
+    metadata = self.get_metadata()
+
+    if self.is_genesis:
+        metadata.closest_block = self._settings.GENESIS_BLOCK_HASH
+        return
+
+    closest_block: Block | None = None
+    dependency_ids = self.parents + [tx_input.tx_id for tx_input in self.inputs]
+
+    for vertex_id in dependency_ids:
+        vertex = self.storage.get_transaction(vertex_id)
+        vertex_meta = vertex.get_metadata()
+        this_closest_block: Block
+
+        if isinstance(vertex, Block):
+            assert vertex_meta.closest_block is None
+            this_closest_block = vertex
+        elif isinstance(vertex, Transaction):
+            this_closest_block_id = (
+                self._settings.GENESIS_BLOCK_HASH if vertex.is_genesis else not_none(vertex_meta.closest_block)
+            )
+            this_closest_block = self.storage.get_block(this_closest_block_id)
+        else:
+            raise NotImplementedError
+
+        if not closest_block or (this_closest_block.get_height() > closest_block.get_height()):
+            closest_block = this_closest_block
+
+    assert closest_block is not None
+    assert closest_block.hash is not None
+    metadata.closest_block = closest_block.hash
+```
+
+A migration will also be necessary to populate this new metadata for existing transactions.
+
+## Example - Releasing new Nano Contracts
+
+To illustrate the usage of Feature Activation for Transactions, we'll demonstrate how it would be used to release a new Nano Contract.
+
+To release a new Nano Contract, a new value would be added to the `TxVersion` enum, allowing deserialization of a different kind of transaction. Since the `FeatureService.is_feature_active_for_transaction()` method requires a `Transaction` instance, it cannot be accessed during deserialization, as the instance hasn't been construct yet. The deserialization will always succeed, even if the Feature fails to become `ACTIVE`.
+
+Then, a new method must be added in the verification phase of `BaseTransaction`, to assert that the deserialized `TxVersion` is valid. In that method, `is_feature_active_for_transaction()` would be called, and depending on the state returned, the `TxVersion` representing the new Nano Contract would be accepted or not.
+
+This demonstrates that there are two possible known use cases for Feature Activation for Transactions:
+
+1. Changing behavior in transaction validation. In this case, the usage is straight forward, simply a call to `FeatureService.is_feature_active_for_transaction()` to check whether the new feature is `ACTIVE`.
+2. Changing behavior in transaction deserialization. In this case, it may be necessary to create a new transaction validation method that verifies the validity of using a new deserialization feature.
+
+## Mutability of transaction parents
```

Comment:
Same here: move to the Guide-level explanation.

Reply:
I don't think this should be moved to the guide-level, this is just a consideration on whether the solution has this problem or not.